pkg/executor/sortexec/topn_chunk_heap.go

```diff
@@ -0,0 +1,170 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sortexec
+
+import (
+	"container/heap"
+	"fmt"
+
+	"github.com/pingcap/tidb/pkg/executor/internal/exec"
+	"github.com/pingcap/tidb/pkg/types"
+	"github.com/pingcap/tidb/pkg/util/chunk"
+	"github.com/pingcap/tidb/pkg/util/memory"
+)
+
+// topNChunkHeap implements heap.Interface.
+type topNChunkHeap struct {
+	compareRow func(chunk.Row, chunk.Row) int
+	greaterRow func(chunk.Row, chunk.Row) bool
+
+	// rowChunks is the chunks to store row values.
+	rowChunks *chunk.List
+	// rowPointer store the chunk index and row index for each row.
+	rowPtrs []chunk.RowPtr
+
+	isInitialized bool
+	isRowPtrsInit bool
+
+	memTracker *memory.Tracker
+
+	totalLimit uint64
+	idx        int
+
+	droppedRowNum     int64
+	fieldTypes        []*types.FieldType
+	droppedRowsString string
+}
+
+func (h *topNChunkHeap) init(topnExec *TopNExec, memTracker *memory.Tracker, totalLimit uint64, idx int, greaterRow func(chunk.Row, chunk.Row) bool, fieldTypes []*types.FieldType) {
+	h.memTracker = memTracker
+
+	h.rowChunks = chunk.NewList(exec.RetTypes(topnExec), topnExec.InitCap(), topnExec.MaxChunkSize())
+	h.rowChunks.GetMemTracker().AttachTo(h.memTracker)
+	h.rowChunks.GetMemTracker().SetLabel(memory.LabelForRowChunks)
+
+	h.compareRow = topnExec.compareRow
+	h.greaterRow = greaterRow
+
+	h.totalLimit = totalLimit
+	h.idx = idx
+	h.isInitialized = true
+
+	h.fieldTypes = fieldTypes
+}
+
+func (h *topNChunkHeap) initPtrs() {
+	h.memTracker.Consume(int64(chunk.RowPtrSize * h.rowChunks.Len()))
+	h.initPtrsImpl()
+}
+
+func (h *topNChunkHeap) initPtrsImpl() {
+	h.rowPtrs = make([]chunk.RowPtr, 0, h.rowChunks.Len())
+	for chkIdx := 0; chkIdx < h.rowChunks.NumChunks(); chkIdx++ {
+		rowChk := h.rowChunks.GetChunk(chkIdx)
+		for rowIdx := 0; rowIdx < rowChk.NumRows(); rowIdx++ {
+			h.rowPtrs = append(h.rowPtrs, chunk.RowPtr{ChkIdx: uint32(chkIdx), RowIdx: uint32(rowIdx)})
+		}
+	}
+	h.isRowPtrsInit = true
+}
+
+func (h *topNChunkHeap) clear() {
+	h.rowChunks.Clear()
+	h.memTracker.Consume(int64(-chunk.RowPtrSize * len(h.rowPtrs)))
+	h.rowPtrs = nil
+	h.isRowPtrsInit = false
+	h.isInitialized = false
+	h.idx = 0
+}
+
+func (h *topNChunkHeap) update(heapMaxRow chunk.Row, newRow chunk.Row) {
+	if h.greaterRow(heapMaxRow, newRow) {
+		// Evict heap max, keep the next row.
+		h.rowPtrs[0] = h.rowChunks.AppendRow(newRow)
+		heap.Fix(h, 0)
+		h.droppedRowsString = fmt.Sprintf("%s; %s", h.droppedRowsString, heapMaxRow.ToString(h.fieldTypes))
+	} else {
+		h.droppedRowsString = fmt.Sprintf("%s; %s", h.droppedRowsString, newRow.ToString(h.fieldTypes))
+	}
+	h.droppedRowNum++
+}
+
+func (h *topNChunkHeap) processChkNoSpill(chk *chunk.Chunk) {
+	for i := 0; i < chk.NumRows(); i++ {
+		heapMaxRow := h.rowChunks.GetRow(h.rowPtrs[0])
+		newRow := chk.GetRow(i)
+		h.update(heapMaxRow, newRow)
+	}
+}
+
+func (h *topNChunkHeap) processChkWithSpill(chk *chunk.Chunk) {
```

Comment:
What does "WithSpill" mean here?

Reply:
> What does "WithSpill" mean here?

There is no difference between `processChkWithSpill` and `processChkNoSpill`. I will rename them to `processChk`